README
modsec2sguil
============
Written by Victor Julien <victor@inliniac.net>


What is it?
-----------
It is a set of Perl scripts that enables you to add ModSecurity alerts to
go into the Sguil NSM system. It acts as an agent to Sguil.


Required software
-----------------
- Apache2
- ModSecurity 2 (1.9.4 may work but is untested), tested up to 2.6.0.
- Sguil 0.8.0 server
- Perl
- IO::Socket::SSL library (on Debian/Ubuntu libio-socket-ssl-perl package)


Install/Usage
-------------

- Create a user to run the agent as -

Add a user with the following command:
# useradd -u 400 -d /home/sguil sguil

This creates a user called sguil. The specific uid is given so transfering
files between different sensors is easier. It is not required. Choose an
uid that is not already in use.


- Installation of files - 

You can place the two binary files (the .pl ones) anywhere you want. Since
I like having all files together I'm putting them in /nsm/bin. Place the
modsec_agent.conf in /nsm/etc

The two libraries (the .pm files) should be in /usr/lib/perl5/ (or use
a symlink for that.


- Agent configuration -

Modsec2sguil uses a configuation file that uses the same syntax as Sguil
agents. Take the example configution file and adapt it to your needs. It
is advisable to enable debugging and disable daemon mode while testing.

Set the RUNAS variable to the user you created above.

This doc assumes the configfile is placed in /nsm/etc


- Apache2 setup -

The following asumes your Apache2 logs to /var/log/apache2/ and
that ModSecurity2 concurrent logging will log to /var/log/apache2/audit_log/data/

Step 1. Creating a queue directory.

modsec_queue.pl adds the events to the queue directory so modsec_agent.pl can
process them in the right order. The directory can be anywhere but note that
when you intend to run modsec_agent.pl under reduced privileges (which is
highly recommended) the directory and it's parents need to be owned by the
user you want to run the agent as.

Create the directory /nsm/queue and set the correct permissions on it:
chown sguil /nsm -R

Step 2. Setup the Apache2/ModSecurity2 config

Put the following lines in your Apache2/ModSecurity2 configuration:

SecAuditLogType Concurrent
SecAuditLogStorageDir /var/log/apache2/audit_log/data/
SecAuditLog "|/path/to/modsec_queue.pl -c /nsm/etc/modsec_agent.conf /var/log/apache2/audit_log/data/ /var/log/apache2/audit_log/index"
SecAuditLogParts ABCDEFGHZ

NOTE: the SecAuditLog line changed from previous version of modsec2sguil!

Restart apache2. You should see event files appear in the queue directoy.
Depending on your setup they might appear on simple webserver visits or
you might have to run a tool like nikto to trigger events.

Look at you apache2 error log for any errors.


- Putting it together -

Next, use modsec_agent.pl to connect to Sguil. It is run as follows:
modsec_agent.pl -c /nsm/etc/modsec_agent.conf

The script enters an endless loop in which it will continuesly check for new
alert files in the queue dir. Press Ctrl-C to kill it. If everything works
fine, you can enable daemon mode.


Support
-------
Mail me at <victor@inliniac.net> or hop in #snort-gui at
irc.freenode.net. I'm using the nick VictorJ.


Future plans / known limitations
-------------------------------
Better error checking.
Enable a way to log when running in daemon mode.
More docs.
Think of a name for this little project ;-)


Special thanks to
-----------------
Ivan Ristic of Breach Security for releasing modsec-auditlog-collector.pl as GPL.
And for creating ModSecurity of course!

Bamm Visscher for answering my endless questions. Of course, Sguil rocks,
so thanks for that as well.



ChangeLog
ChangeLog for Modsec2sguil

Version 0.8

- Send/expect 0.8.0 identifier from/to server.

Version 0.8-dev8

- Fix parsing of lines with multiple 'tag' entries.

Version 0.8-dev7 March 26th,2008

- Update protocol string to reflect Sguil 0.7.0 stable release
- Support ModSecurity 2.5.x, thanks for the addition Ryan Cummings

Version 0.8-dev6 September 18th, 2007

- Catch errors in syslog so they will be non-fatal.
- Fix broken syslog level.

Version 0.8-dev5 August 19th, 2007

- Add syslog logging.
- Add option to log all transactions, not just alerts.

Version 0.8-dev4 August 16th, 2007

- Fix a bug where the agent would exit if the connection to the server was
  lost.

Version 0.8-dev3 August 16th, 2007

- Make sure alerts get prio 1-4, non-alerts prio 5.
- Clean up SguilAgent.pm rtgenevent function.
- Fix rule id being part of the event.
- Add option to send all http events to Sguil, even non-alerts.

Version 0.8-dev2 August 14th, 2007

- Fix missing RUNAS van causing a error message.
- Add an option IGNORE_HTTP_CODES to the config, to optionally no treat
  certain codes as alerts, and thus not send them to Sguil.
- Fix PINGs sending duplicate \r\n

Version 0.8-dev1 August 13th, 2007

- Converted into a real agent for Sguil 0.7 (no more barnyard replacement)
- Agent can drop privileges
- Agent can daemonize
- Pinging the server is supported
- The agent reconnects to the server if the connection is lost
- Agent supports SSL for the connection to the server
- A sguil-compatible configuration file is now used
- A debug mode was added

Version 0.7 March 18th, 2007

- add support for ModSecurity 2.x alerts.
- fix wrong severity to prio conversion.



ModsecAlert.pm
#
# Copyright (C) 2006-2007 by Victor Julien <victor@inliniac.net>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the
# Free Software Foundation, Inc.,
# 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
#

#
# Object for parsing ModSecurity autitlog event files.
#

my $version = "0.8";
package ModsecAlert;

use fields qw (	store );

sub new {
    my $type = shift;
    my ModsecAlert $self = fields::new(ref $type || $type);

    $self->{store} = undef;

    return $self;
}

sub version {
    my ModsecAlert $self = shift;

    return $version;
}

#
# Returns a COPY of the alert hash
#
sub getalerthash {
    my ModsecAlert $self = shift;

    return $self->{store};
}

#
# Takes the location of a file as argument, 
# opens it, parses it and stores it into
# the $self->{store} hash.
#
sub parsefile {
    my ModsecAlert $self = shift;
    my $filename = shift;

    my $line_num = 1;
    my $fileid;
    my $section = "-";
    my %event;
    my @reqhd; # request headers
    my @rsphd; # response headers

    # initialize some vars in case the alert doesn't contain them
    $self->{store}{"file"} = "";
    $self->{store}{"themsg"} = "";
    $self->{store}{"http_request_body"} = "";
    $self->{store}{"severity"} = "DEBUG"; # lowest by modsec
    $self->{store}{"rev"} = 0;
    $self->{store}{"http_request_urlencoded"} = 0;

    $self->{store}{"http_request_protocol"} = "";
    $self->{store}{"http_response_protocol"} = "";

    open (FILE , $filename) or
        die "Could not open file $filename: $!\n";

    while (<FILE>)
    {
        # store all lines of the file so we can use it
        # as the alert payload later
        $self->{store}{"file"} = $self->{store}{"file"} . $_;

        # Get this file's unique id
        if ($line_num == 1)
        {
            my @parsed_line = /^--(\S+)-A--$/;
            if ( @parsed_line == 0 ) {
                goto error;
            } else {
                #print "parsed line @parsed_line\n";

                ( $fileid ) = @parsed_line;
                #print "fileid " . $fileid . "\n";
            }
        }	
        # parse the section character
        if ( /^--$fileid-\S+--$/ ) {
            my @parsed_line = /^--$fileid-(\S+)--$/;
            ( $section ) = @parsed_line;
            #print "section " . $section . "\n";
        }
        #section A -- time, ipaddresses and ports
        elsif ($section eq "A")
        {
            if (m/^\[.*\] .*$/) {
                my @parsed_line = /^\[(\d+)\/(\S+)\/(\d+):(.*) (.*)\] (\S+) (\S+) (\d+) (\S+) (\d+)$/;
                if ( @parsed_line == 0 ) {
                    goto error;
                } else {
                    ( $self->{store}{"day"}, $self->{store}{"monthstr"}, $self->{store}{"year"},
                      $self->{store}{"timestr"}, $self->{store}{"tz"}, $self->{store}{"uniqueid"},
                      $self->{store}{"sipstr"}, $self->{store}{"sp"}, $self->{store}{"dipstr"},
                      $self->{store}{"dp"} ) = @parsed_line;

                      # all further processing is done only if we know this
                      # is an alert. PreprocessAlert processes it further.
                }
            }
        }
        #section B request headers
        elsif ( $section eq "B" )
        {
            my @header_list;

            if ( m/^\S+ .* HTTP\/\d\.\d$/ )
            {
                my @parsed_line = /^(\S+) (.*) (\S+)$/;
                if ( @parsed_line == 0 ) {
                    goto error;
                } else {
                    ( $self->{store}{"http_request_method"} , $self->{store}{"http_request_uri"} , $self->{store}{"http_request_protocol"} ) = @parsed_line;
                }
            } elsif ( m/^\S+\: .*$/ ) {
                my @parsed_line = /^(\S+)\: (.*)$/;
                if ( @parsed_line == 0 ) {
                    goto error;
                } else {
                    my %header;

                    ( $header{"name"} , $header{"value"} ) = @parsed_line;
                    push(@reqhd, \%header);

                    # A bit hackish, but this saves us walking the list
                    # for a second time later
                    if ( ($header{"name"} eq "Content-Type" ) &&
                         ($header{"value"} eq "application/x-www-form-urlencoded")) {
                        $self->{store}{"http_request_urlencoded"} = 1;
                    }
                }
            }
            # handle "GET /"
            elsif ( m/^\S+ .*/ )
            {
                my @parsed_line = /^(\S+) (.*)$/;
                if ( @parsed_line == 0 ) {
                    goto error;
                } else {
                    ( $self->{store}{"http_request_method"} , $self->{store}{"http_request_uri"} ) = @parsed_line;
                }
            }
            # Store the list in our object
            $self->{store}{"http_request_headers"} = \@reqhd;

            #print "DEBUG: headers ".$self->{store}{"http_request_headers"}." ".@reqhd."\n";
        }
        #section C request body
        elsif ( $section eq "C" )
        {
            $self->{"store"}{"http_request_body"} = $self->{"store"}{"http_request_body"} . $_;
        }
        # section F request response
        # can in some cases be completely empty
        elsif ( $section eq "F" )
        {
            if ( m/^HTTP\/\d\.\d \d+ .*/ )
            {
                my @parsed_line = /^(\S+) (\d+) (.*)$/;
                if ( @parsed_line == 0 ) {
                    goto error;
                } else {
                    ( $self->{store}{"http_response_protocol"} , $self->{store}{"code"} , $self->{store}{"http_response_reason_phrase"} ) = @parsed_line;
                }
            }
        }
        #section H, alert info, mod_sec response
        elsif ( $section eq "H" )
        {
            # ModSecurity 1.9.4
            if ( m/^Message\: Access denied with code \d+\. .*/ )
            {
                my @parsed_line = /^Message\: Access denied with code (\d+)\. (.*) \[severity \"(\S+)\"\]$/;
                if ( @parsed_line == 0 ) {
                    goto error;
                } else {
                    ( $self->{store}{"code"} , $self->{store}{"themsg"} , $self->{store}{"severity"} ) = @parsed_line;
                }
            }
            # Modsecurity 2.5 adds a tag at the end of the line now for classification,
            # Thanks to Ryan Cummings.
            elsif ( m/^Message\: Access denied with code \d+ \(phase \d\)\. .*\[severity \"\S+\"\] (\[tag \"(\S+)\"\])?/ )
            {
                my @parsed_line = /^Message\: Access denied with code (\d+) \(phase \d\)\. (.*) \[severity \"(\S+)\"\] (\[tag \"(\S+)\"\])?.*$/;
                if ( @parsed_line == 0 ) {
                    goto error;
                } else {
                    ( $self->{store}{"code"} , $self->{store}{"themsg"} , $self->{"store"}{"severity"} ) = @parsed_line;
                }
            }
            # ModSecurity 2.x -- adds phase, and no longer adds severity
            # but sometimes it does:
            # example: Message: Access denied with code 403 (phase 2). Unconditional match in SecAction. [msg "LOCAL comment spam: forbidden word"] [severity "DEBUG"]
            # So look for a line with severity first
            elsif ( m/^Message\: Access denied with code \d+ \(phase \d\)\. .*\[severity \"\S+\"\]/ )
            {
                my @parsed_line = /^Message\: Access denied with code (\d+) \(phase \d\)\. (.*) \[severity \"(\S+)\"\].*$/;
                if ( @parsed_line == 0 ) {
                    goto error;
                } else {
                    ( $self->{store}{"code"} , $self->{store}{"themsg"} , $self->{"store"}{"severity"} ) = @parsed_line;
                }
            }
            elsif ( m/^Message\: Access denied with code \d+ \(phase \d\)\. .*/ )
            {
                my @parsed_line = /^Message\: Access denied with code (\d+) \(phase \d\)\. (.*)$/;
                if ( @parsed_line == 0 ) {
                    goto error;
                } else {
                    ( $self->{store}{"code"} , $self->{store}{"themsg"} ) = @parsed_line;
                    $self->{"store"}{"severity"} = 0;
                }
            }
            elsif ( m/^Message\: Access denied with redirect to.*/ )
            {
                my @parsed_line = /^Message\: Access denied with redirect to \[(.*)\]\. (.*) \[severity \"(\S+)\"\]$/;
                if ( @parsed_line == 0 ) {
                    goto error;
                } else {
                   ( $self->{store}{"redirecturl"} , $self->{store}{"themsg"} , $self->{store}{"severity"} ) = @parsed_line;
                }
            }
            # Apache-Error: [file "/build/buildd/apache2-2.0.55/build-tree/apache2/modules/aaa/mod_access.c"] [line 282] [level 3] client denied by server configuration: /var/www/cgi-bin
            elsif ( m/^Apache-Error\: \[file .*\] .* \[level \d+\] .*/ )
            {
                #print "ModsecAlert.pm: Apache Error!\n";
                my @parsed_line = /^Apache-Error\: \[file .*\] .* \[level \d+\] (.*)$/;
                if ( @parsed_line == 0 ) {
                    goto error;
                } else {
                    ( $self->{store}{"themsg"} ) = @parsed_line;
                    $self->{store}{"themsg"} = "Apache-Error: " . $self->{store}{"themsg"};
                }
                #print "msg \"" . $self->{store}{"themsg"} . "\"\n";
            }

            # if we do not yet have a code (e.g. in a redirect case)
            # we try to find it here
            if (not defined($self->{store}{"code"}) )
            {
                if (m/^Action\: Intercepted \((\d+)\).*/)
                {
                    my @parsed_line = /^Action\: Intercepted \((\d+)\)$/;
                    if ( @parsed_line == 0 ) {
                        goto error;
                    } else {
                        ( my $code ) = @parsed_line;
                        #print "Action " . $code . "\n";
                        $self->{store}{"code"} = $code;
                    }
                }
            }
        }#sections done
        $line_num++;
    }

    close(FILE);
    return;

    # a bit ugly, but it saves us from duplicating a lot of code:
    # close file, chomp, die with error message.
error:
    close(FILE);
    chomp();
    die "Error parsing line $line_num: $_";
}

1;



modsec_agent.pl
#!/usr/bin/perl -w

#
# Copyright (C) 2006-2011 by Victor Julien <victor@inliniac.net>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the
# Free Software Foundation, Inc.,
# 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
#

# Agent for feeding ModSecurity alerts to the Sguil NSM system.
# 
# Reads files or symlinks from the queue directory, feeds them
# to Sguil and removes the file/symlink.
#

use strict;
use warnings;
use diagnostics;

use IO::Socket;
use DirHandle;
use Getopt::Std;
use Time::Local;
use POSIX 'setsid';

use ModsecAlert;
use SguilAgent;

use strict "vars";
use strict "subs";

my $version = "0.8";
my $license = "GNU GPLv2. See http://www.gnu.org/licenses/gpl2.txt for more information";

#
# CONVERSION FUNCTIONS
#

sub ConvertASCIItoHEX
{
    my $asciistr = shift @_;

    my $hexstr = join '', unpack "H*", $asciistr;
    #print $hexstr . "\n";

    return($hexstr);
}

#
# This doesn't return 5, the lowest for Sguil because we use that for
# non-alert events
#
sub ConvertSEVERITYtoPRIO
{
    my $severity = shift @_;

    my %sev = ( "EMERGENCY" => 1,
                "ALERT"     => 1,
                "CRITICAL"  => 2,
                "ERROR"     => 2,
                "WARNING"   => 3,
                "NOTICE"    => 3,
                "INFO"      => 4,
                "DEBUG"     => 4 );

    if (not defined $sev{$severity}) {
        return 4; # default to 4 which is the lowest we use for alerts
    }

    return $sev{$severity};
}

sub ConvertMONTHSTRtoDEC
{
    my $monthstr = shift;

    my %mon = ( "Jan" => 1, "Feb" => 2, "Mar" => 3, "Apr" => 4,
                "May" => 5, "Jun" => 6, "Jul" => 7, "Aug" => 8,
                "Sep" => 9, "Oct" => 10,"Nov" => 11,"Dec" => 12 );

    if (not defined $mon{$monthstr}) {
        return -1;
    }

    return $mon{$monthstr};
}

sub ConvertIPtoDEC
{
    my $ip = shift;
        
    if (not $ip =~ /^\d+\.\d+\.\d+\.\d+$/) {
        return(0);
    }

    ( my $one , my $two , my $three , my $four ) = split(/\./, $ip);
    my $dec = $one * 16777216 + $two * 65536 + $three * 256 + $four;

    return($dec);
}


#
# ALERT PROCESSING
#

#
# Some preprocessing before sending the alert
# to Sguil.
#
# This function is only called for actual alerts.
#
# Returns:  -1 on error
#       0 on success
#
sub PreprocessAlert
{
    # get the alert hash from the caller (by reference)
    my $ref = shift @_;

    #
    # SECTION A
    #

    # stupid trick to convert month(str) to month(dec)
    my $month = ConvertMONTHSTRtoDEC $ref->{"monthstr"};
    if ($month != -1) {
        $ref->{"month"} = $month;
    } else {
        print "Month parse failed.\n";
        return -1;
    }

    # convert sipstr and dipstr to dec as well
    $ref->{"sipdec"} = ConvertIPtoDEC $ref->{"sipstr"};
    $ref->{"dipdec"} = ConvertIPtoDEC $ref->{"dipstr"};

    # timezone handling: the time/date is logged like this:
    # 13/Aug/2006:21:59:24 +0200, which is 19:59.24 UTC. In
    # Sguil we want the UTC value, so we convert it here.
    # 
    # first parse the timestring
    (my $hour, my $min, my $sec ) = split(/:/, $ref->{"timestr"});

    # get the unixtime in gm, month is from 0 to 11.
    my $time = timegm($sec,$min,$hour,$ref->{"day"},$ref->{"month"}-1,$ref->{"year"});

    # parse the timezone offset. It looks like +0200 or -0100
    $_ = $ref->{"tz"};
    my @parse = /(.{1})(\d{2})(\d{2})$/;
    if (@parse == 0) {
        print "Parsing timezone information failed.\n";
        return -1;
    } else {
        (my $op, my $tz_hour, my $tz_min) = @parse;
        my $tz_val = $tz_hour * 3600 + $tz_min * 60;

        # +0200 means we have to subtrackt 2 hrs to get UTC
        if ( $op eq "+" ) {
            $time -= $tz_val;
        } elsif ( $op eq "-" ) {
            $time += $tz_val;
        } else {
           print "Unknown operator $op in timezone string " . $ref->{"tz"} . "\n";
           return -1;
        }

        #print "time after tz apply: $time, gmtime " . gmtime($time) . "\n";

        # convert the new unix time
        # Fri Aug 18 04:35:35 2006
        # Mon Aug  7 07:51:23 2006 <= note two spaces, split(/ / won't work
        # that is what the \ ? takes care of.
        # 
        # update day, timestr and year
        $_ = gmtime($time);
        @parse = /(\S*) (\S*) \ ?(\d*) (.*) (\d+)/;
        if ( @parse == 0 ) {
            print "Date/time parsing error, parsing $_ failed\n";
            return -1;
        } else {
            (my $dayname, $ref->{"monthstr"}, $ref->{"day"}, $ref->{"timestr"}, $ref->{"year"} ) = @parse;
        }

        # update month
        $month = ConvertMONTHSTRtoDEC $ref->{"monthstr"};
        if ($month != -1) {
            $ref->{"month"} = $month;
        } else {
           print "Month parse failed, parsing " . $ref->{"monthstr"} . ".\n";
           return -1;
        }

        #print $ref->{"timestr"} . "\n";
    }

    # add a padding zero to match the barnyard behaviour.
    if ( $ref->{"day"} < 10 ) {
        $ref->{"day"} = "0" . $ref->{"day"};
    }
    if ( $ref->{"month"} < 10 ) {
        $ref->{"month"} = "0" . $ref->{"month"};
    }

    $ref->{"time"} = $ref->{"year"} . "-" . $ref->{"month"} . "-" . $ref->{"day"} . " " . $ref->{"timestr"};

    #
    # SECTION H
    #

    # check for a message with a message, get it?
    if( defined( $ref->{"themsg"} ) ) {
        #print "themsg: " .  $ref->{"themsg"} . "\n";
        ($ref->{"rev"})    = ($ref->{"themsg"} =~ /.*\[rev \"(\d+)\"\].*/);
        ($ref->{"id"})     = ($ref->{"themsg"} =~ /.*\[id \"(\d+)\"\].*/);
        ($ref->{"themsg"}) = ($ref->{"themsg"} =~ /.*\[msg \"(.*)\"\].*/);

        #if (defined $ref->{"rev"}) { 
        #    print "REV " . $ref->{"rev"} . "\n";
        #}
        #if (defined $ref->{"themsg"}) { 
        #    print "MSG " . $ref->{"themsg"} . "\n";
        #}
        #if (defined $ref->{"id"}) { 
        #    print "ID  " . $ref->{"id"} . "\n";
        #}
    }

    # lets see what kind of class we are going to use
    # we use *-attack for stuff that is blocked, *-activity
    # for the rest
    if ($ref->{"code"} < 400) {
        $ref->{"class"} = "web-application-activity";
        $ref->{"prio"} = 5; # lowest prio
    } else {
        $ref->{"class"} = "web-application-attack";
        $ref->{"prio"} = ConvertSEVERITYtoPRIO $ref->{"severity"};
    }

    # Event message
    if ((not defined $ref->{"themsg"}) || ($ref->{"themsg"} eq "")) {
        $ref->{"themsg"} = $ref->{"http_response_reason_phrase"};
    }
    $ref->{"msg"} = "MSc " . $ref->{"code"} . " " . $ref->{"themsg"};

    # hex the message
    $ref->{"hexmsg"} = ConvertASCIItoHEX $ref->{"msg"};

    # hex the payload
    $ref->{"payload"} = ConvertASCIItoHEX $ref->{"file"};

    return 0;
}

# tries to drop privs to $string
# $string can be either "user" or "user:group"
sub RunAs {
    my $string = shift @_;

    my $user;
    my $group;

    ($user, $group) = split(/:/, $string, 2);

    if (not defined $user) {
        return;
    }
    if ($user eq "") {
        return;
    }

    my $curname = getpwuid($<);
    my $curuid  = getpwnam($curname);

    my $newuid = 0;
    eval { $newuid = getpwnam($user) };
    if ($@) {
        die $@;
    } elsif (not defined $newuid) {
        die "getpwnam for $user failed";
    }

    if ($curuid == $newuid) {
        return;
    }

    my $newgid = $(;
    if (defined $group && $group ne "") {
        $newgid = getgrnam($group);
        if (not defined $newgid) {
            die "getgrnam for $group failed";
        }
    }

    # set the new uid
    $( = $) = $newgid;
    $< = $> = $newuid;

    return;
}

sub Daemonize
{
    $| = 1;

    chdir '/'                 or die "Can't chdir to /: $!";
    open STDIN, '/dev/null'   or die "Can't read /dev/null: $!";
    open STDOUT, '>/dev/null' or die "Can't write to /dev/null: $!";
    defined(my $pid = fork)   or die "Can't fork: $!";
    exit if $pid;
    setsid                    or die "Can't start a new session: $!";
    open STDERR, '>&STDOUT'   or die "Can't dup stdout: $!";
}


#
# Returns a list of filenames of event files.
# Returns 0 if no files were found.
#
sub GetAlertFiles {
    my $dir = shift;
    #print $dir . "\n";

    my $dh = DirHandle->new($dir) or die "can't open dir $dir: $!\n";
    return sort
           grep { /^$dir\/modsec\.log\..*/ } # we want only the files that start with modsec.log.
           map  { "$dir/$_" }
           grep { !/^\./    }
           $dh->read();
}

#
# START MAIN
#

my $sguil = new SguilAgent;

# global vars
my %g_ignore_http_codes = ();
my $g_handle_all_events = 0;
my $g_event_dir;
my $g_runas = "";
my $g_debug = 0;
my $g_daemon = 0;
my $g_symlink = 0;
my $g_symlink_rm_orig = 0;

# fill the g_ignore_http_codes hash
sub GetIgnoreCodes {
    my $codes;
    my @list;
    eval { $codes = $sguil->getvar ("IGNORE_HTTP_CODES") };
    if ($@ || $codes eq "") {
        return;
    }

    @list = split (/ /, $codes);
    foreach my $code (@list) {
        $g_ignore_http_codes{$code} = 1;
    }
}

# returns 1 if the code is in the ignore list
# returns 0 otherwise
sub TestIgnoreCode {
    my $code = shift;

    if (not defined $g_ignore_http_codes{$code}) {
        return 0;
    }

    return 1;
}

sub process_events {
    my $event_dir = shift @_;

    while ( 1 )
    {
        my @list = GetAlertFiles ( $event_dir );
        if ( @list == 0 ){
            # only print these when debug is 2 to prevent flooding the logs
            if ($g_debug == 2) {
                $sguil->debuglog ("No eventfiles, sleeping for a second." );
            }
            $sguil->ping();
            sleep 1;
        }
        else
        {
            # parse each file and check if it really was an alert
            foreach my $file (@list) {
                $sguil->debuglog ( "Eventfile found: $file." );
                my $valid = 0;

                if (-l $file) {
                    my $real_file = readlink($file);
                    $sguil->debuglog ( "Symlink $file to $real_file." );
                    if (-r $real_file) {
                        $valid = 1;
                    } else {
                        $sguil->debuglog ( "real file doesn't exist" );
                    }
                } else {
                    if(-r $file) {
                        $valid = 1;
                    } else {
                        $sguil->debuglog ( "file doesn't exist" );
                    }
                }

                if ($valid == 1) {
                    # get a ModsecAlert object
                    my $alert = new ModsecAlert;

                    # parse the alert file
                    eval { $alert->parsefile( $file ) };
                    if ($@) {
                        die $@;
                    }

                    # get a copy from the hash of the alert
                    my %event_hash = %{$alert->getalerthash()};
                    # create a reference to work with
                    my $hash_ref = \%event_hash;

                    # Determine if we will continue processing the event.
                    #
                    if (exists($event_hash{"code"}) &&
                            ($event_hash{"code"} ge 400 || $g_handle_all_events == 1) &&
                            TestIgnoreCode($event_hash{"code"}) == 0)
                    {
                        # preprocess the data so it becomes what Sguil expects
                        if ( PreprocessAlert ( $hash_ref ) == 0 ) {
                            $sguil->debuglog ( "Processing Event \"" . $event_hash{"themsg"} . "\"" );

                            # send the event as a GenericEvent to Sguild
                            eval { $sguil->rtgenevent( $hash_ref ) };
                            if ($@) {
                                die $@;
                            }

                            # increase the cid so the next event gets it's uniq cid.
                            $sguil->incrcid();

                            $sguil->debuglog ( "Event accepted by server." );
                        }
                    }
                } else {
                    print "removing invalid file\n";
                }

                if ($g_symlink == 1 && $g_symlink_rm_orig == 1) {
                    my $orig_file = readlink($file);
                    if (defined $orig_file && $orig_file ne "") {
                        $sguil->debuglog ( "Original file to remove: $orig_file" );

                        my $rem = unlink ($orig_file);
                        if ($rem != 1 ) {
                            $sguil->warninglog ( "warning could not remove $orig_file: $!" );
                        }
                    }
                }

                my $removed = unlink ($file);
                if ($removed != 1 ) {
                    # fatal error, because we keep processing the same
                    # event over and over again if it wasn't removed.
                    $sguil->errorlog ( "FATAL ERROR: could not remove $file: $!" );
                    exit 1;
                }
            }
        }
    }
}


#
# option parsing
# 
my %option = ();
getopts("vc:", \%option);

if ( defined $option{v} ) {
    my $modsec = new ModsecAlert;
    print "\n";
    print "modsec_agent.pl version $version (ModsecAlert.pm " . $modsec->version() . ", SguilAgent.pm " . $sguil->version() . ").\n\n";
    print "Copyright (c) 2006-2011 by Victor Julien <victor\@inliniac.net>.\n";
    print "Released under $license.\n";
    print "\n";
    exit 0;
}

# Load the configution file.
#
eval { $sguil->loadcnf($option{c}) };
if ($@) {
    print "FATAL ERROR: loading configuration file \"" . $option{c} . "\" failed: $@\n";
    exit 1;
}

# Get from the config what we need right here.
#

# See if we want to drop privs and if so, do it
# asap.
eval { $g_runas = $sguil->getvar("RUNAS") };
eval { RunAs($g_runas) };
if ($@) {
    print "FATAL ERROR: dropping privs failed: $@\n";
    exit 1;
}

# we require EVENT_DIR for picking up events.
eval { $g_event_dir = $sguil->getvar ("EVENT_DIR") };
if ($@ || $g_event_dir eq "") {
    print "FATAL ERROR: EVENT_DIR variable missing or invalid, please review your configuration file.\n";
    exit 1;
}


eval { $g_symlink = $sguil->getvar ("SYMLINK") };
if ($@ || $g_symlink eq "") {
    $g_symlink = 0;
}
eval { $g_symlink_rm_orig = $sguil->getvar ("SYMLINK_REMOVE_ORIGINAL") };
if ($@ || $g_symlink_rm_orig eq "") {
    $g_symlink_rm_orig = 0;
}
eval { $g_handle_all_events = $sguil->getvar ("HANDLE_ALL_EVENTS") };
if ($@ || $g_handle_all_events eq "") {
    $g_handle_all_events = 0;
}
eval { $g_debug = $sguil->getvar ("DEBUG") };
if ($@ || $g_debug eq "") {
    $g_debug = 0;
}
GetIgnoreCodes();
eval { $g_daemon = $sguil->getvar("DAEMON") };
if ($@ || $g_daemon ne 1) {
    $g_daemon = 0;
}

#
# Okay, let do daemon if we are configured to.
#
if ( $g_daemon == 1 ) {
    Daemonize();
}

#
# Main processing loop. We connect, process events, reconnect.
#
while (1) {
    my $done = 0;

    # Try to connect. Retry every 5 seconds until it succeeds.
    while ($done == 0) {
        eval { $sguil->connect () };
        if ($@) { 
            $sguil->warninglog ("Setting up the connection to the server failed: $@");
            sleep(5);
        } else {
            $done = 1;
        }
    }
    $sguil->debuglog ( "Connected to server" );

    eval { process_events($g_event_dir) };
    if ($@) {
        $sguil->errorlog ( "processing events failed: $@" );
    }

    eval { $sguil->disconnect() };
    if ($@) {
        $sguil->errorlog ( "FATAL ERROR: closing connection problem: $@" );
        exit 1;
    }
}

exit 0;

